orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/persistence/jedis/AbstractRedisExecutionRepository.java

```diff
@@ -0,0 +1,343 @@
+
+import static com.netflix.spinnaker.orca.pipeline.model.Execution.ExecutionType.PIPELINE;
+import static java.lang.System.currentTimeMillis;
+import static java.lang.String.format;
+
+@Slf4j
+abstract class AbstractRedisExecutionRepository implements ExecutionRepository {
+
+  static final TypeReference<List<Task>> LIST_OF_TASKS =
+    new TypeReference<List<Task>>() {
+    };
+  static final TypeReference<Map<String, Object>> MAP_STRING_TO_OBJECT =
+    new TypeReference<Map<String, Object>>() {
+    };
+  final RedisClientDelegate redisClientDelegate;
+  final Optional<RedisClientDelegate> previousRedisClientDelegate;
+  protected final ObjectMapper mapper = OrcaObjectMapper.newInstance();
+  final int chunkSize;
+  final Scheduler queryAllScheduler;
+  final Scheduler queryByAppScheduler;
+  private final Registry registry;
+
+  AbstractRedisExecutionRepository(
+    Registry registry,
+    RedisClientDelegate redisClientDelegate,
+    Optional<RedisClientDelegate> previousRedisClientDelegate,
+    Scheduler queryAllScheduler,
+    Scheduler queryByAppScheduler,
+    Integer threadPoolChunkSize
+  ) {
+    this.redisClientDelegate = redisClientDelegate;
+    this.previousRedisClientDelegate = previousRedisClientDelegate;
+    this.queryAllScheduler = queryAllScheduler;
+    this.queryByAppScheduler = queryByAppScheduler;
+    this.chunkSize = threadPoolChunkSize;
+    this.registry = registry;
+  }
+
+  @Override
+  public void cancel(@Nonnull String id) {
+    cancel(id, null, null);
+  }
+
+  @Override
+  public void cancel(@Nonnull String id, String user, String reason) {
+    String key = fetchKey(id);
+
+    getRedisDelegateForId(key).withCommandsClient(c -> {
+      HashMap<String, String> data = new HashMap<>();
+      data.put("canceled", "true");
+      if (StringUtils.isNotEmpty(user)) {
+        data.put("canceledBy", user);
+      }
+      if (StringUtils.isNotEmpty(reason)) {
+        data.put("cancellationReason", reason);
+      }
+      ExecutionStatus currentStatus = ExecutionStatus.valueOf(c.hget(key, "status"));
+      if (currentStatus == ExecutionStatus.NOT_STARTED) {
+        data.put("status", ExecutionStatus.CANCELED.name());
+      }
+      c.hmset(key, data);
+    });
+  }
+
+  @Override
+  public boolean isCanceled(@Nonnull String id) {
+    String key = fetchKey(id);
+
+    return getRedisDelegateForId(key).withCommandsClient(c -> {
+      return Boolean.valueOf(c.hget(key, "canceled"));
+    });
+  }
+
+  @Override
+  public void pause(@Nonnull String id, String user) {
+    String key = fetchKey(id);
+
+    getRedisDelegateForId(key).withCommandsClient(c -> {
+      ExecutionStatus currentStatus = ExecutionStatus.valueOf(c.hget(key, "status"));
+      if (currentStatus != ExecutionStatus.RUNNING) {
+        throw new IllegalStateException(format(
+          "Unable to pause pipeline that is not RUNNING (executionId: %s, currentStatus: %s)",
+          id,
+          currentStatus.toString()
+        ));
```

Comment:
In the spirit of trying to make more specific exception classes, any feels on a new exception class? `UnpausablePipelineException extends IllegalStateException` or similar?